### ISSUE TITLE
post processor Vsphere not escaping slashes for target VM for ovftool command

### ISSUE THREAD

**Issue by packernewb** [hashicorp/packer#7368]
Hello packer team! This is my first post ever, so please let me know if I missed any critical information for bug reproduction. Thanks!

My environment: 
Packer Version = 1.3.1
Windows 10 x64d

I am using VMware 14 Workstation installed locally, the ovftool CLI (VMware ovftool 4.3.0 (build-7948156)), my datacenter is VCSA with an ESXI host.

My plan is to clone a generic template 2012r2 windows VM, run some basic scripts on it, then upload it to our VCSA. I can build using vmware-vmx, I can provision just fine, I have a problem with the post-processor vsphere.

My claim is that packer is not handling escaping slashes in the target VM for the command created and executed for the ovftool. The reason I say this is because if I execute the command in the very last line of the log in ovftool with two fast changes, it works just fine. The changes are specifically: I replace <password> with my URI encoded password and I add an additional slash in front of any in the path to the vmx file like this:

Original:
`ovftool --acceptAllEulas --name=packer-clone-1551714952 --datastore=datastore2 --noSSLVerify=true --diskMode=thin --vmFolder="//" --network=VLAN111_Group output\packer-created-1551714952\packer-created-1551714952.vmx vi://packerman:<password>@<correctVcenter+esxihost>
`

Modified and works:
`ovftool --acceptAllEulas --name=packer-clone-1551714952 --datastore=datastore2 --noSSLVerify=true --diskMode=thin --vmFolder="//" --network=VLAN111_Group output\\packer-created-1551714952\\packer-created-1551714952.vmx vi://packerman:<REAL_PASSWWORD>@<correctVcenter+esxihost>
`

`output\ vs output\\`

The current behavior I am seeing is that packer just freezes on this ovftool command. 30min plus, so I interrupt it. I assume this is wrong because if I use ovftool it takes max five minutes to complete upload. Perhaps there is another issue of any errors reported by the ovftool not getting into the packerlog and getting stuck in execution.

This is the last line that gets executed based on the log output.

https://github.com/hashicorp/packer/blob/2a39dad5bc0b5c9d3fdc24c1e6bdcda8b59c8887/post-processor/vsphere/post-processor.go#L161

Build file:
https://gist.github.com/packernewb/445b13c08f918f583b580d707201741f

Log file:
https://gist.github.com/packernewb/ba6049bf1f3ab50d0ac7f27dc897b8a8

-packernewb

Edit: fixed links to gist

**Comment by SwampDragons** [comment]
In v1.3.3 I did make a change fixing how we escape usernames; I'd like you to give that a try just to make sure that this hasn't already been fixed.

If that doesn't work, try out this binary
[packer.zip](https://github.com/hashicorp/packer/files/2928636/packer.zip)
 where I'm escaping the slashes in the source path. If this is really the root issue I'll be pretty surprised because I would have thought we'd know about a pathing issue this dramatic; if this is really the case than the vsphere post-processor is pretty much broken on all windows builds.

**Comment by packernewb** [comment]
Ok, so after testing today i'm still not sure where this leaves me. Turns out I was testing several things at the same time. I eliminated more fronts of failure today.

First:  I was using Git Bash for Windows. turns out that my expectation that to target the file test in directory output should be:
`output\\test.vmx`

when this works in cmd:
`output\test.vmx` (which is what packers creates).

So when I see the ovftool command in the log file, it was working if I escape the slashes in Bash, or if I just run the command in CMD - depending on how Packer executes this it could be fine, but it would be helpful to know how so when we describe directory paths in our config files we can be consistent with the way packer will execute and look for those targets. Example: For this to work, I needed to change the "vm_folder" in my config from `"\"//\"" to "\"/\""` so that it shows up in the ovftool command as `--vmFolder="/"`

I also learned that you need to have the user authenticating against the VCSA, not the ESXI specifically, not this issue related but I thought I would share for future thread searchers. I was failing sometimes because of failure to authenticate because my service account on ESXI was different than VCSA.

I tried the same build as before with packer 1.3.3 and the packer 1.2.6 you gave me specifically.

**packer 1.3.3** - doesn't freeze, but immediately errors with no explanation. All I get is:

`2019/03/05 15:49:39 packer.exe: 2019/03/05 15:49:39 Starting ovftool with parameters: --acceptAllEulas --name=packer-clone-1551818779 --datastore=datastore2 --noSSLVerify=true --diskMode=thin --vmFolder="/" --network=VLAN111_Group output\packer-created-1551818778\packer-created-1551818778.vmx vi://packerman%40vsphere.local:<password>@<correctDatacenter,correctURL,correctESXIhost>`

`2019/03/05 15:49:39 [INFO] (telemetry) ending vsphere
2019/03/05 15:49:39 Deleting original artifact for build 'vmware-vmx'
2019/03/05 15:49:39 ui error: Build 'vmware-vmx' errored: 1 error(s) occurred:`

`\* Post-processor failed: Error uploading virtual machine: exit status 1`

`2019/03/05 15:49:39 Builds completed. Waiting on interrupt barrier...
2019/03/05 15:49:39 machine readable: error-count []string{"1"}
2019/03/05 15:49:39 ui error: 
==> Some builds didn't complete successfully and had errors:
2019/03/05 15:49:39 machine readable: vmware-vmx,error []string{"1 error(s) occurred:\n\n* Post-processor failed: Error uploading virtual machine: exit status 1\n\n"}
2019/03/05 15:49:39 ui error: --> vmware-vmx: 1 error(s) occurred:`

`\* Post-processor failed: Error uploading virtual machine: exit status 1`

`2019/03/05 15:49:39 ui: 
==> Builds finished but no artifacts were created.
2019/03/05 15:49:39 [INFO] (telemetry) Finalizing.
2019/03/05 15:49:39 waiting for all plugin processes to complete...`

**packer 1.2.6** - can't validate because i'm using "esxi_host" in my configuration file which that version I suppose does not understand.

So Packer might be working fine with regards to the slashes, however it is still not successful uploading the VM to VCSA with my configuration, nor does the error code help at all to figure out why. Yet the same command to ovftool does work in CMD.

**Comment by SwampDragons** [comment]
Oh weird I did not mean to build off of 1.2.6. Sorry about that. Here's a build off master with the slash escape added.  [packer.zip](https://github.com/hashicorp/packer/files/2937792/packer.zip)


The revelation about you using a special command line is pretty important; it explains why I haven't seen someone running packer via powershell or CMD raise this issue. 

I'm not sure Packer can directly support your exporting via the vsphere postprocessor if you've got the custom command prompt; I'm not at all versed in git bash for windows, but I'm not sure how we'd introspect to figure out your custom shell and modify our escapes. Theoretically we could add a flag, but that gets messy from a maintenance perspective.

My recommendation here for a short-term workaround would probably be to call the shell-local post-processor to use ovftool directly, rather than depending on Packer to handle the special-case pathing right with its more generic post-processor.

**Comment by SwampDragons** [comment]
I'm going to close this as "wontfix" for now because of the custom shell being used, but if you haven't found a workaround please reach out to the mailing list and we'll help you figure something out.

**Comment by packernewb** [comment]
@SwampDragons 

> Mailing list

Which one is that?

**Comment by SwampDragons** [comment]
ah, sorry -- https://groups.google.com/forum/#!forum/packer-tool
